File: backend/app/domain/metrics.py
"""Метрики использования рекомендаций для панели руководителя (FR10, UC-07) —
порт metrics.ts. Считаются по логу обратной связи (FR8) и составу рекомендаций."""

from __future__ import annotations

from typing import Any

from .clock import js_round, parse_ms


def _rate(part: int, whole: int) -> int:
    if whole == 0:
        return 0
    return js_round((part / whole) * 100)


def compute_supervisor_metrics(
    recommendations: list[dict[str, Any]],
    feedback: list[dict[str, Any]],
    managers: list[dict[str, Any]],
) -> dict[str, Any]:
    # Отклонение черновика CRM — не решение по рекомендации, в метрики не входит.
    feedback = [e for e in feedback if e.get("kind") != "crm_draft"]

    # Последнее решение по каждой рекомендации.
    latest: dict[str, dict[str, Any]] = {}
    for event in feedback:
        prev = latest.get(event["recommendationId"])
        if not prev or parse_ms(event["createdAtIso"]) > parse_ms(prev["createdAtIso"]):
            latest[event["recommendationId"]] = event

    decisions = list(latest.values())
    decision_counts = {
        "accepted": sum(1 for d in decisions if d["decision"] == "accepted"),
        "edited": sum(1 for d in decisions if d["decision"] == "edited"),
        "rejected": sum(1 for d in decisions if d["decision"] == "rejected"),
    }

    total = len(recommendations)
    decided = len(decisions)
    accepted = decision_counts["accepted"]
    quality_count = decision_counts["accepted"] + decision_counts["edited"]

    high_recs = [r for r in recommendations if r["level"] == "high"]
    high_covered = sum(1 for r in high_recs if r["recommendationId"] in latest)

    recent = sorted(feedback, key=lambda e: parse_ms(e["createdAtIso"]), reverse=True)[:6]

    by_manager: list[dict[str, Any]] = []
    for m in managers:
        rec_ids = {r["recommendationId"] for r in recommendations if r["managerId"] == m["id"]}
        recs_total = len(rec_ids)
        mgr_decisions = [d for d in decisions if d["recommendationId"] in rec_ids]
        acc = sum(1 for d in mgr_decisions if d["decision"] == "accepted")
        edt = sum(1 for d in mgr_decisions if d["decision"] == "edited")
        rej = sum(1 for d in mgr_decisions if d["decision"] == "rejected")
        dec = len(mgr_decisions)
        by_manager.append(
            {
                "managerId": m["id"],
                "managerName": m["fullName"],
                "total": recs_total,
                "decided": dec,
                "accepted": acc,
                "edited": edt,
                "rejected": rej,
                "usageRate": _rate(dec, recs_total),
                "acceptanceRate": _rate(acc, dec),
            }
        )

    return {
        "totalRecommendations": total,
        "decided": decided,
        "usageRate": _rate(decided, total),
        "acceptanceRate": _rate(accepted, decided),
        "qualityRate": _rate(quality_count, decided),
        "decisionCounts": decision_counts,
        "highPriorityTotal": len(high_recs),
        "highPriorityCovered": high_covered,
        "coverageRate": _rate(high_covered, len(high_recs)),
        "recentDecisions": recent,
        "byManager": by_manager,
        "preliminary": decided < 8,
    }

File: src/domain/metrics.ts
// ============================================================================
// Метрики использования рекомендаций для панели руководителя (FR10, UC-07).
// Считаются по логу обратной связи (FR8) и составу рекомендаций.
// Все метрики при малом объёме данных помечаются как предварительные.
// ============================================================================

import type { FeedbackDecision, FeedbackEvent, PriorityLevel } from "./types";

export interface RecommendationRef {
  recommendationId: string;
  level: PriorityLevel;
  managerId: string;
}

export interface ManagerBreakdown {
  managerId: string;
  managerName: string;
  total: number;
  decided: number;
  accepted: number;
  edited: number;
  rejected: number;
  usageRate: number;
  acceptanceRate: number;
}

export interface SupervisorMetrics {
  totalRecommendations: number;
  decided: number;
  usageRate: number; // доля рекомендаций с принятым решением
  acceptanceRate: number; // доля «принято» среди решённых
  qualityRate: number; // доля «принято или отредактировано» среди решённых
  decisionCounts: Record<FeedbackDecision, number>;
  highPriorityTotal: number;
  highPriorityCovered: number;
  coverageRate: number;
  recentDecisions: FeedbackEvent[];
  byManager: ManagerBreakdown[];
  preliminary: boolean;
}

function rate(part: number, whole: number): number {
  if (whole === 0) return 0;
  return Math.round((part / whole) * 100);
}

export function computeSupervisorMetrics(input: {
  recommendations: RecommendationRef[];
  feedback: FeedbackEvent[];
  managers: { id: string; fullName: string }[];
}): SupervisorMetrics {
  const { recommendations, managers } = input;
  // Отклонение черновика CRM — не решение по рекомендации, в метрики не входит.
  const feedback = input.feedback.filter((e) => e.kind !== "crm_draft");

  // Последнее решение по каждой рекомендации.
  const latestByRec = new Map<string, FeedbackEvent>();
  for (const event of feedback) {
    const prev = latestByRec.get(event.recommendationId);
    if (!prev || new Date(event.createdAtIso) > new Date(prev.createdAtIso)) {
      latestByRec.set(event.recommendationId, event);
    }
  }

  const decisions = [...latestByRec.values()];
  const decisionCounts: Record<FeedbackDecision, number> = {
    accepted: decisions.filter((d) => d.decision === "accepted").length,
    edited: decisions.filter((d) => d.decision === "edited").length,
    rejected: decisions.filter((d) => d.decision === "rejected").length,
  };

  const total = recommendations.length;
  const decided = decisions.length;
  const accepted = decisionCounts.accepted;
  const qualityCount = decisionCounts.accepted + decisionCounts.edited;

  // Покрытие приоритетных кейсов: доля high-приоритетных рекомендаций,
  // по которым есть решение менеджера.
  const highRecs = recommendations.filter((r) => r.level === "high");
  const highCovered = highRecs.filter((r) => latestByRec.has(r.recommendationId)).length;

  const recentDecisions = [...feedback]
    .sort((a, b) => new Date(b.createdAtIso).getTime() - new Date(a.createdAtIso).getTime())
    .slice(0, 6);

  const byManager: ManagerBreakdown[] = managers.map((m) => {
    const recs = recommendations.filter((r) => r.managerId === m.id);
    const recIds = new Set(recs.map((r) => r.recommendationId));
    const mgrDecisions = decisions.filter((d) => recIds.has(d.recommendationId));
    const acc = mgrDecisions.filter((d) => d.decision === "accepted").length;
    const edt = mgrDecisions.filter((d) => d.decision === "edited").length;
    const rej = mgrDecisions.filter((d) => d.decision === "rejected").length;
    const dec = mgrDecisions.length;
    return {
      managerId: m.id,
      managerName: m.fullName,
      total: recs.length,
      decided: dec,
      accepted: acc,
      edited: edt,
      rejected: rej,
      usageRate: rate(dec, recs.length),
      acceptanceRate: rate(acc, dec),
    };
  });

  return {
    totalRecommendations: total,
    decided,
    usageRate: rate(decided, total),
    acceptanceRate: rate(accepted, decided),
    qualityRate: rate(qualityCount, decided),
    decisionCounts,
    highPriorityTotal: highRecs.length,
    highPriorityCovered: highCovered,
    coverageRate: rate(highCovered, highRecs.length),
    recentDecisions,
    byManager,
    preliminary: decided < 8,
  };
}
